sway-core/src/ir_generation/function.rs

```diff
@@ -1719,30 +1843,39 @@ impl<'eng> FnCompiler<'eng> {
             )
             .add_metadatum(context, cond_span_md_idx);
 
-        let return_type = convert_resolved_typeid_no_span(
-            self.engines.te(),
-            self.engines.de(),
-            context,
-            &return_type,
-        )
-        .unwrap_or_else(|_| Type::get_unit(context));
         let merge_block = self.function.create_block(context, None);
         // Add a single argument to merge_block that merges true_value and false_value.
-        // Rely on the type of the ast node when creating that argument
-        let merge_val_arg_idx = merge_block.new_arg(context, return_type);
-        if !true_block_end.is_terminated(context) {
-            true_block_end
-                .append(context)
-                .branch(merge_block, vec![true_value]);
-        }
-        if !false_block_end.is_terminated(context) {
-            false_block_end
-                .append(context)
-                .branch(merge_block, vec![false_value]);
-        }
-
-        self.current_block = merge_block;
-        Ok(merge_block.get_arg(context, merge_val_arg_idx).unwrap())
+        // Rely on the type of the ast node when creating that argument.
+        let val = if true_value.is_terminator && false_value.is_terminator {
+            // Corner case: If both branches diverge, then the return type is 'Unknown', which we can't
```

Comment:
I like how simply this corner case is handled! In my original thought of returning `DivergingValue` enum the result would still be the merge block argument, but the enclosing enum variant would carry the information about divergence. This is much nicer!